### PR TITLE
feat(core): add API to provide CSP nonce for inline stylesheets

### LINKED ISSUE #6361: Inline <style> elements violates style-src Content Security Policy
See https://developer.mozilla.org/en-US/docs/Web/Security/CSP/CSP_policy_directives#style-src

The framework should support users that want to build their apps for CSP. In this case, the `style-src` directive would be in violation by Angular 2's use of inline `<style>` elements for things like CSS encapsulation.

I was curious if Angular 1 supported CSP, and it seems that Angular 1 indeed generates a stylesheet (`build/angular-csp.css`) that users can consume for CSP mode. 

cc @tbosch @matsko

### PR DESCRIPTION
**Pull request by crisbeto**
Angular uses inline styles to insert the styles associated with a component. This violates the strict styles [Content Security Policy](https://web.dev/strict-csp/) which doesn't allow inline styles by default. One way to allow the styles to be applied is to set a `nonce` attribute on them, but because the code for inserting the stylesheets is deep inside the framework, users weren't able to provide it without accessing private APIs.

These changes add a new `CSP_NONCE` injection token that will allow users to provide a nonce, if their app is using CSP. If the token isn't provided, the framework will look for an `ngCspNonce` attribute on the app's root node instead. The latter approach is provided as a convenience for apps that render the `index.html` through a server, e.g. `<app ngCspNonce="{% randomNonceAddedByTheServer %}"></app>`.

This PR addresses adding the nonce to framework-generated styles. There will be follow-up PRs that add support for it in critical CSS tags in the CLI, and in Angular Material.

Fixes #6361.

### CODE REVIEW

File: packages/platform-server/src/styles_host.ts [hunk 1]
```diff
@@ -7,26 +7,34 @@
...
 @Injectable()
 export class ServerStylesHost extends SharedStylesHost {
   private head: any = null;
   private _styleNodes = new Set<HTMLElement>();
+  private nonce = inject(CSP_NONCE, {optional: true});
 
   constructor(@Inject(DOCUMENT) doc: any, @Inject(APP_ID) private appId: string) {
     super();
-    this.head = doc.getElementsByTagName('head')[0];
+    this.head = doc.head || doc.getElementsByTagName('head')[0];
```
Comment (crisbeto) [inline packages/platform-server/src/styles_host.ts:21]: Unrelated change, I just saw that we could get the `head` directly without querying the DOM. I left the old approach as a fallback in case it was there as a workaround for Domino.

File: packages/platform-browser/src/dom/dom_renderer.ts [hunk 2]
```diff
@@ -91,7 +91,8 @@ export class DomRendererFactory2 implements RendererFactory2, OnDestroy {
   constructor(
       private eventManager: EventManager, private sharedStylesHost: DomSharedStylesHost,
       @Inject(APP_ID) private appId: string,
-      @Inject(REMOVE_STYLES_ON_COMPONENT_DESTROY) private removeStylesOnCompDestory: boolean) {
+      @Inject(REMOVE_STYLES_ON_COMPONENT_DESTROY) private removeStylesOnCompDestory: boolean,
+      @Inject(CSP_NONCE) @Optional() private nonce?: string|null) {
```
Comment (crisbeto) [inline packages/platform-browser/src/dom/dom_renderer.ts:95]: This uses `@Inject` instead of `inject`, because we had a bunch of tests that were constructing classes manually which lacked a DI context.

**Review by jessicajaniuk — APPROVED** [review]
Aside from @AndrewKushnir's comment, this looks good to me

reviewed-for: fw-core, fw-platform-server, public-api

**Review by dylhunn — APPROVED** [review]
reviewed-for: public-api

**Review by alxhub — APPROVED** [review]
Reviewed-for: public-api

File: packages/platform-browser/src/dom/shared_styles_host.ts [hunk 3]
```diff
@@ -149,6 +150,11 @@ export class SharedStylesHost implements OnDestroy {
   private addStyleToHost(host: Node, style: string): void {
     const styleEl = this.getStyleElement(host, style);
 
+    if (this.nonce) {
```
Comment (alan-agius4) [inline packages/platform-browser/src/dom/shared_styles_host.ts:153]: Would we great if you can move this https://github.com/angular/angular/blob/7831a7a95f48c3942b7b63307c1df683c9c52f3d/packages/platform-browser/src/dom/shared_styles_host.ts#L144 where all the style tag manipulation happen.

File: packages/platform-browser/src/dom/shared_styles_host.ts [hunk 4]
```diff
@@ -149,6 +150,11 @@ export class SharedStylesHost implements OnDestroy {
...
+      // Uses a keyed write to avoid issues with property minification.
```
Comment (alan-agius4) [inline packages/platform-browser/src/dom/shared_styles_host.ts:154]: OOC, does closure minify this property?

Comment (crisbeto) [reply packages/platform-browser/src/dom/shared_styles_host.ts]: I'm not sure, but I wanted to keep it safe since I've seen it minify things like `el.style.foo` before.

**Comment by pkozlowski-opensource** [comment]
This PR was merged into the repository by commit 17e9862653758ebdbd29771cd4ec8a59436497d6.
